Initial documentation

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes

### Description

Closes #6

gsteel committed:
Initial documentation for the Filters

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on docs/book/filters/to-e164.md:
````diff
@@ -0,0 +1,66 @@
+echo $filter->filter('44 (0) 1234 567 890'); // "+441234567890"
+```
+
+## Changing the default country code
````
```suggestion
## Changing the Default Country Code
```

froschdesign commented on docs/book/filters/to-e164.md:
````diff
@@ -0,0 +1,66 @@
+$filter->setOptions($options);
+```
+
+## Filtering invalid data
````
For headlines the Chicago Manual of Style capitalization rules must be used:

1. Capitalize the first and the last word.
1. Capitalize nouns, pronouns, adjectives, verbs, adverbs, and subordinate conjunctions.
1. Lowercase articles (a, an, the), coordinating conjunctions, and prepositions.
1. Lowercase the ‘to’ in an infinitive (I want to play guitar).

The online tool [capitalizemytitle.com](https://capitalizemytitle.com/style/Chicago/) can help to capitalize and case convert the headlines.

```suggestion
## Filtering Invalid Data
```

froschdesign commented on docs/book/filters/to-international-phone-number.md:
````diff
@@ -0,0 +1,66 @@
+echo $filter->filter('44 (0) 1234 567 890'); // "+44 1234 567890"
+```
+
+## Changing the default country code
````
```suggestion
## Changing the Default Country Code
```

froschdesign commented on docs/book/filters/to-international-phone-number.md:
````diff
@@ -0,0 +1,66 @@
+$filter->setOptions($options);
+```
+
+## Filtering invalid data
````
```suggestion
## Filtering Invalid Data
```

froschdesign commented on docs/book/filters/to-national-phone-number.md:
````diff
@@ -0,0 +1,66 @@
+echo $filter->filter('+441234567890'); // "01234 567890"
+```
+
+## Changing the default country code
````
```suggestion
## Changing the Default Country Code
```

froschdesign commented on docs/book/filters/to-national-phone-number.md:
````diff
@@ -0,0 +1,66 @@
+$filter->setOptions($options);
+```
+
+## Filtering invalid data
````
```suggestion
## Filtering Invalid Data
```

gsteel committed:
Change headings to title-case

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: George Steel <george@netglue.uk>

gsteel committed:
Initial documentation for the validator

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on docs/book/validators/phone-number.md:
````diff
@@ -0,0 +1,143 @@
+
+## Basic Usage
+
+```php
+use Laminas\I18n\PhoneNumber\Validator\PhoneNumber;
+
+$validator = new PhoneNumber();
+$validator->isValid('+4930123456'); // true
+```
````
The most code examples can be reduced.

```suggestion
```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->isValid('+4930123456'); // true
```
```

froschdesign commented on docs/book/validators/phone-number.md:
````diff
@@ -0,0 +1,143 @@
+The validator is capable of validating a range of number types; mobile numbers, fixed lines, premium rate and emergency numbers to name a few.
+By default, the validator will accept any number type:
+
+```php
+$validator = new PhoneNumber([
+    'country' => 'US',
+]);
+$validator->isValid('911'); // true
+```
````
```suggestion
```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber([
    'country' => 'US',
]);
$validator->isValid('911'); // true
```
```

froschdesign commented on docs/book/validators/phone-number.md:
````diff
@@ -0,0 +1,143 @@
+];
+
+$validator = new PhoneNumber($options);
+// Options can also be changed at runtime with:
+$validator->setOptions($options);
+```
````
This should be extracted to a separate code example with description:

``````suggestion
```

Options can also be changed at runtime with:

```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->setOptions($options);
```
``````

froschdesign commented on docs/book/validators/phone-number.md:
````diff
@@ -0,0 +1,143 @@
+
+Each of the 3 options have companion "setters" to change the runtime behaviour of the validator after it has been constructed:
+
+```php
+use Laminas\I18n\PhoneNumber\PhoneNumberValue;
+use Laminas\I18n\PhoneNumber\Validator\PhoneNumber;
+
+$validator = new PhoneNumber();
+
+$validator->setCountry('US'); // Set the default country to US
+$validator->setCountry('de_DE'); // A locale string can also be used to set the country
+
+$validator->setCountryContext('my-country-input'); // Set the country validation context key
+
+$validator->setAllowedTypes(PhoneNumberValue::TYPE_VOIP); // Set the acceptable number types
+```
````
Longer descriptions does not belong in code example. Headlines should be used to break longer text sections to brighten up.

Example:

``````suggestion
### Set Country Code

```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->setCountry('US');
```

A locale string can also be used to set the country.

```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->setCountry('de_DE');
```

### Set the Country Validation Context Key

```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->setCountryContext('my-country-input');
```

### Set the Acceptable Number Types

```php
$validator = new Laminas\I18n\PhoneNumber\Validator\PhoneNumber();
$validator->setAllowedTypes(
    Laminas\I18n\PhoneNumber\PhoneNumberValue::TYPE_VOIP
);
```
``````

gsteel committed:
Initial documentation for the phone number form input

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Shorten code blocks and extract comments to discrete sections

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: George Steel <george@netglue.uk>

froschdesign commented on docs/book/element/phone-number.md:
```diff
@@ -0,0 +1,94 @@
+
+$phone = new PhoneNumber('number');
+$phone->setLabel('Phone Number');
+$phone->setAttributes([
+    'autocomplete' => 'tel', // 'tel-national' is also an option.
+]);
```
Is the autocomplete option recommended? Otherwise please add this part to the callout where the autocomplete attribute is described.

froschdesign commented on docs/book/element/phone-number.md:
```diff
@@ -0,0 +1,94 @@
+TIP: **Autocomplete Attribute**
+To make it easier for users to autofill valid information, the [autocomplete attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) can be used. A value of `tel` should autofill the users phone number with the country dialling code, whereas a value of `tel-national` will autofill the number without the country dialling code.
+
+## Available Options
```
I am sorry, because I forgot to say that the current page structure of the form elements is bad and should not be used.

froschdesign commented on docs/book/element/phone-number.md:
````diff
@@ -0,0 +1,94 @@
+]);
+```
+
+## Public Methods
+
+The following methods are specific to the `PhoneNumber` element; all other methods
+defined by the [parent `Element` class](https://docs.laminas.dev/laminas-form/v3/element/element/#public-methods) are also
+available.
+
+Method signature                    | Description
+----------------------------------- | -----------
+`getInputSpecification() : array`   | See write-up below.
+`setOptions(array $options) : void` | Set options for an element of type `PhoneNumber`. The accepted options, in addition to the inherited options of [`Laminas\Form\Element`](https://docs.laminas.dev/laminas-form/v3/element/element/#public-methods) are `default_country`, which calls `setDefaultCountry()`, `country_context`, which calls `setCountryContext()` and `allowed_types`, which calls `setAl ...
+`setDefaultCountry(string $code): void`  | Sets the default country to validate national phone numbers against.
+`setCountryContext(string $inputName): void`              | Sets the name of another input in your form that can be used to determine the country for validating national phone numbers, instead of the default country.
+`setAllowedTypes(int $types): void` | Is a bitmask of phone number types to limit what type of number are considered valid.
+
+`getInputSpecification()` returns an input filter specification, which includes the `StringTrim` filter, and a phone number validator configured with the `default_country`, `country_context`, and `allowed_types` options.
+
+The options are described in greater detail in the documentation for the [phone number validator](../validators/phone-number.md).
````
The same here, this table with the different methods is unsuitable for explanations of this type.
Each method should be described in a separate section with a headline, description and code example.

gsteel replied:
I've reformatted this section with H3, copy and a code example for each method. Can you please check it's looking OK now?

gsteel committed:
Adds a code example to autocomplete attribute callout

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-work the 'Public Methods' section to prose instead of a table

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add code examples for each public method with arguments

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix doc linting error

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on docs/book/element/phone-number.md:
````diff
@@ -0,0 +1,135 @@
+])
+```
+
+### `setDefaultCountry(string $code): void`
````
If we add the code signature here, it means the same as if I read the code itself.

In the description for the validators we have already good headlines, like:

- Set Country Code
- Set the Country Validation Context Key
- Set the Acceptable Number Types

froschdesign commented on docs/book/element/phone-number.md:
```diff
@@ -0,0 +1,135 @@
+
+### `setOptions(array $options): void`
+
+Set options for an element of type `PhoneNumber`. The accepted options, in addition to the inherited options of [`Laminas\Form\Element`](https://docs.laminas.dev/laminas-form/v3/element/element/) are `default_country`, which calls `setDefaultCountry()`, `country_context`, which calls `setCountryContext()` and `allowed_types`, which calls `setAllowedTypes()`.
```
Lists should be used for all kinds of enumerations, they offer the following advantages:

- optical loosening
- set accents
- avoids tapeworm enumerations

Therefore, I think this is a better overview:

```suggestion
Set options for an element of type `PhoneNumber`.
The accepted options, in addition to the inherited options of [`Laminas\Form\Element`](https://docs.laminas.dev/laminas-form/v3/element/element/) are:

- `default_country`, which calls `setDefaultCountry()` 
- `country_context`, which calls `setCountryContext()`
- `allowed_types`, which calls `setAllowedTypes()`.
```

froschdesign commented on docs/book/element/phone-number.md:
```diff
@@ -0,0 +1,135 @@
+defined by the [parent `Element` class](https://docs.laminas.dev/laminas-form/v3/element/element/#public-methods) are also
+available.
+
+### `getInputSpecification(): array`
+
+Returns an [input filter specification](https://docs.laminas.dev/laminas-inputfilter/specs/), which includes the `StringTrim` filter, and a phone number validator configured with the `default_country`, `country_context`, and `allowed_types` options.
+The validator options are described in greater detail in the documentation for the [phone number validator](../validators/phone-number.md).
```
I like this part a lot because this is the first time that the predefined specification is broken down exactly.
However, I would place this at the bottom of the page, because an user will not interact further with the method itself.

froschdesign reviewed (changes requested):
Minor changes, but looks good! 👍🏻

gsteel committed:
Improve readability with a list

Co-authored-by: Frank Brückner <info@froschdesignstudio.de>
Signed-off-by: George Steel <george@netglue.uk>

gsteel committed:
Improvements to element documentation

- Use headings instead of method signatures
- Moves `getInputSpecification` description to the end of the document

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Minor linting fix

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds initial documentation for the `phoneNumberFormat` view helper

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
Thanks for the continued reviews, improvements and help @froschdesign - Just added docs for the view helper and I think that should be everything then?

froschdesign reviewed (approved):
Thank you for your effort and patience! 👍🏻

froschdesign commented:
@gsteel 
> Just added docs for the view helper and I think that should be everything then?

I'm sure there will be a few small additions, but I'll take them on when we add the pages to the documentation of laminas-i18n.
